README.md
# README #

Steps to get this repository up and running and developers information.

Release notes are in a separate .md file

### What is this repository for? ###

* Dexter - Qt C++ fully featured readout and control software for the Medipix3 RX chip

### How do I get set up? ###

* Use the dexter-setup.sh BASH script.
    * ```./dexter-setup.sh``` for user only setup
    * ```./dexter-setup.sh -m dev``` for the full developer setup
* OpenGL libraries are required, these are typically installed as part of a graphics card driver. This means a GPU or integrated GPU is required.
    * If you cannot run ```glxgears```, you will not be able to run Dexter
* Example build command:
```
qmake SPIDR-MPX3.pro -spec linux-g++ && make qmake_all && make -j8
cd Release
./Dexter
```

### Relevant developers ###

* Lead - Bram Bouwens
* Support - Kiavash Matin
* Support - Navrit Bal

-------------------------------------------------------------

# Developers notes (only)
Last change: 27-09-2016
Author: Amber van Keeken (primary) & Navrit Bal (notes)

## Setting up a new view (eg. Configuration, DACs etc.) using QtCreator

1. The 'view' are QWidgets that are part of a StackedWidget. In QtCreator, go to the design form of the main window and right-click on the stackedWidget -> Insert page. A new QWidget is then added to the stackedWidget.
2. Create a new Qt Designer Form Class with QWidget as its base class.
3. Promote the QWidget to the newly made class.
- I can’t see the newly made class in mpx3gui.ui? (designer view)
    - Right click on the new class, *Promote to ...*
    - Add your new class by name under QWidget (the header file is auto completed for you below
    - Press *Add* then *Promote*
    - Add new class and header to mpx3gui.h
    - Add to new header in public
        Ui::CLASSNAME * GetUI(){ return ui; };
        void SetMpx3GUI(Mpx3GUI *p);
    - Add in private
        Ui::CLASSNAME *ui;
        Mpx3GUI * _mpx3gui;
    - Include your <ui_HEADERNAME.h>
    - **NOTE: THE ORDER OF YOUR TAB IN THE STACKED WIDGET GIVES IT AN INDEX NUMBER which affects things like keyboard shortcuts.** You cannot search for this. Use the Qt Designer to change the order to match what the ...id variables say.

4. Add functionalities to the newly made class (widget) as you would for a normal widget, dialog or window.

- Make sure to have a private variable *_mpx3gui* (or something similar) and a public setter (_setMpx3GUI_) that sets the private variable to the main GUI, to have access to the other objects, in particular _dataset_.

1. Add an Action to the menubar in the designer of Qt to mpx3gui.ui. Add an icon to the _icons_ folder in the _SpidrMpx3Eq_ folder.
2. Optional: add a shortcut to the view in mpx3gui.cpp

## Dlib

“_Dlib is a modern C++ toolkit containing
machine learning algorithms and tools for creating complex software
in C++ to solve real world problems.”_
[http://dlib.net/]

This library can be used freely, license:
[http://dlib.net/license.html](http://dlib.net/license.html).

The library is used for its optimization and fast
Fourier transform algorithms. It uses its own data structures, which
have to be used in order to properly use the functions. The Qt data
structures used in the rest of the code are converted to these dlib
structures when a dlib function is needed, through simple for loops.

Simpler and/or better libraries are available for
the optimization and fft, **separately**.
To reduce the number of dependencies, this library was chosen, even
if it means temporarily converting to other (unwieldy) data
structures.

If however having separate libraries may prove to
be superior/preferred, a developer is free to change all this.

### FFT – used in *calcMTFdata()*

The number of rows and columns of the input matrix
to calculate the fast Fourier transform, must be powers of 2. This
means the data used must be (IGNORE? ‘cropped’ in some way or) padded with zeros
to assure this. To do this a dlib function called *order_of_two*
is also used.

### Optimization – used in *fitESFparams()*

For fitting the ESF data, a least squares method
was used. Dlib provides a function that expects its own input vector
and parameter_vector data types. Converting to these data types
requires  small loops. Two prototype functions  (not member functions)
 have to be created: model and residual. The first describes the
function that is to be fitted to the data, the second calculates the
difference between the value that comes out of this model function
and the data, at each data point.

It however does not provide a ‘goodness of fit’ parameter. This has to be calculated separately. For now, the Mean Squared Error and R squared is reported. (χ^2 (chi-squared) is not suitable for a non-linear model).



### R^2 – used in *plotFitESF()*

Dlib provides a function *r_squared* that calculates the R^2, given two vectors: the fitvalues and the datavalues. R^2
is also called the coefficient of determination and ranges from 0–1. 1 being a perfect fit and 0 being a poor fit.

## Calculation of MTF, NPS, DQE:

### Overview DQE calculation:

##### **DQE (n) = NEQ(n) / SNR^2_ideal**

For the DQE we need :

- **NEQ**
**NEQ = G^2 [MTF(n)]^2/ NPS(n)**
for this we need:
    - **(E or P)MTF (n)**(Dobbins, 1995) or (Flynn & Samei, 1999) resp.
    - EMTF can be estimated by averaging several PMTF:
    - P(resampling)MTF is the fourier transform of:
        - **LSF**  
        Line spread function.   
        Derivative of:
            - **ESF**  
            Edge spread function. Pixel value against distance from edge.  
            We need:
                - **Sharp edge image**

    - **NPS (n)**
    for this we need:
        - **Fourier Transform **  
        of RoI’s open beam image(s)
    - **G^2**
    The system gain.

- **SNR^2_ideal**
for this we need:
    - **SNR^2_ideal / exposure**
    a quantity that can be obtained by:
        - Tabulated values for specific beam qualities. (Dobbins, 1995) (Hoeschen, 2001)
        - Estimation by computational modelling. (Flynn & Samei, 1999) See included screen captures.
        They use mass-energy absorption coefficients from: http://physics.nist.gov/PhysRefData/XrayMassCoef/
        In dry air: http://physics.nist.gov/PhysRefData/XrayMassCoef/ComTab/air.html


    - The **exposure**  used in the measurements of MTF and NPS.

### MTF

The user should be given the option to calculate the MTF in different ways.

The way that is implemented works by the following steps:

- Determination of the edge line by finding the point of change from bright to dark or vice versa.
- Calculating the distance to the edge line for each point and plot the value of all points against the distance from the edge.
- Binning these points using a binsize set by the user. This gives the discrete ESF.
- Making a fit through the (binned) datapoints using an error function.
- Differentiation of this fitted function gives the LSF.
    - Differentiation by using the fitting parameters or numerical differentiation are both possible (see options).
- Fast Fourier Transforming of the LSF gives the MTF.
- Maximization of MTF (TO DO, see below).

Another implementation is discussed by Samei et.
al. (1998). This method consists of the following steps:

- Exponential-to-linear transformation.
- Image is converted to an 8-bit binary image by supplying a threshold value (average of the signal of the two sides).
- Determination of angle and position of the edge with double Hough transformation.
- Projection of points to the distance s from the edge. And collection into bins of 0.1*pixelsize. This gives the ESF_k array, the discrete ESF.
- This ESF_k array is smoothed using a 4th order, Gaussian-weighted, moving polynomial fit. Local smoothing does not confine the ESF to a particular mathematical form: for each element in the ESF_k array, a polynomial function is fit using adjacent elements and the initial element value is replaced by the value predicted by the fit. “A least-squares fit is employed for which values near the center-point are strongly weighted by entering a different variance value for each point.” ?? Weighting function is a Gaussian (see form in article), of which parameters were chosen based on a test performed on simulated edge images… (Can we use this then? Do we need to do a similar test?)

- The ESF_k array is numerically differentiated.
- The baseline of the LSF is substracted using a linear fit to the 10-nm-long portions of the LSF tails.
- A Hanning filter with a window width of 20 mm is applied to the LSF to establish a sampling rate of 0.05 cycles/mm.
- Presampling MTF is obtained by a Fast Fourier Transform of the LSF.
- This is repeated for 17 angles in an +-0.16 degrees, by varying by 0.02 degrees each time. The MTF’s are each integrated from 0-2 cycles/mm and the angle associated with the maximum of this MTF integral is identified as the best estimate for the edge angle and its MTF is the result.

The extent of the regions used by Samei are not feasible for the Medipix3… The LSF looks significantly different than ours, so the Hanning filter and baseline substraction or extrapolation of the tails (as in other papers) might not be the way to go. (?)
But we CAN use the iterative MTF maximization by varying the angle. Only one parameter changes and then the entire “Calc MTF” procedure can be repeated:

#### TO DO for MTF calculation:

- Implement slit method.  Try Fujita’s slit method. (“Simple”).
- Maximize MTF by angle variation.
- Get the calculation to work for horizontal edges.
- (?)Better quantization of goodness of fit. -> R^2 and the mean squared error are given.

### NPS calculation

To calculate the NPS, the method as described in
the Handbook of Medical Imaging (Beutel et al.) and by
(Dobbins, 1995).
The method is described by the steps below in order of the
calculation. Which is not necessarily the order of implementation.

#### Steps:

- Determine characteristic curve.
- Convert pixel values to be linearly proportional to exposure (using characteristic curve).
- Take multiple flat field images (at multiple exposures).
- Determine best size of ROI (N) and number of ROI. (M)
- *Correct for background trends from ROI by fitting a planar ramp.*[DONE]
- Calculate the 2D squared Fourier transforms of the ROI. [DONE]
- Take the ensemble average of all these FT of ROIs NPS. [DONE]
- Calculate 1D NPS by using the x=0 and y=0 values and/or averaging a few lines close to the axes (determined by the user).[DONE]

### DQE

According to the method used for the NPS, it is
then not too difficult to calculate the DQE. (see overview)

#### Steps:

- Calculate SNR^2_ideal per exposure from the x-ray spectrum.
- Determine the exposure.
- Calculate NEQ.
- Use EMTF for NEQ calculation (in addition to the PMTF, both should be reported). (Dobbins, 1995)
- Calculate DQE.
- Calculate DQE(0) by extrapolation of the low-frequency part back to zero.

## QCstmDQE class

The user can select a region in the visualization
tab by dragging the mouse with the right mousebutton pressed. An
option window pops up and one can then choose to use the region for
DQE calculation. The GUI then directly switches to the DQEtab, where
the correct data is already imported and plotted. The data is also
binned and plotted. Whether the user sees the unbinned data or not is
determined by the *show prebinning data* checkbox.

When the software is connected to the detector, data can be taken directly from the DQEtab. The view will switch to the visualization tab.

The way the MTF is calculated can be chosen by the radio buttons, checkboxes and lineEdit:

- *Edge vs. Slit radio buttons*
    These let the user choose between the edge and slit method. Both should be
	reported for the most accurate MTF. (TO DO: implement slit method.)
- *Error function vs _Smoothing*
    If *Error function* is selected in the combobox, an error function is fitted to the (binned) data. This then represents the ESF and is used to calculate the derivative (LSF) (see next point). When *Smoothing* is selected, the binned data is smoothed by a 4th order (Gaussian weighted) moving polynomial fit, using a window width as specified by the user.
- *Window width*
    Here the user specifies the window width to be used in the local fitting.
	This must be an uneven number, since there must be as much data
	points to the left of the point to be fitted as to the right. It
	must also be at least 3. And finally the width must be smaller than
	the entire dataset. The program gives off warnings and changes the
	value automatically when this happens.
- *Func der*
When this is checked and an error function has been used (both checked),
	the derivative data is calculated by using the formula for the
	derivative of the error function, using the fitted parameters.
If the error function has not been used, but instead the binned data
	was smoothed, the numerical derivative of this smoothed data is
	taken as the LSF.
    - When not checked, the numerical derivative
		(3-point) of the binned ESF data is taken.  

## OptionsDialog class

The optionsdialog contains the options for the MTF
and NPS calculation all in one place, so as not to clutter the main
interface. Only one optionsdialog exists and is shown
_on_optionsPushbutton_clicked_.
When closed, it is not deleted. The object gets deleted when the
parent (_mpx3gui) is deleted.

Three buttons are available.

- _Ok_
    The settings are applied and saved* and the dialog is closed (but is not
	destroyed). When opening the dialog again, the settings are the way
	the user changed them before.
- _Cancel_
    The settings are not applied and not saved. The changes the user made in
	the ui are changed back to how they were when the user opened the
	window (before any changes were made). The dialog is closed.
- _Apply_
    The settings are applied and saved*, but the dialog is not closed. This
	way the user can see what options are used when calculating the MTF
	and NPS. (These are also noted in the log (see LOG).)

The settings are saved in a private variable of
the optionsDialog_ called currentSettings_.
This is a QHash, which is a container that stores values by a key (in
this case a QString) and a value (in this case an int). This allows
for intuitive looking up and setting of the settings. QHash is faster
than QMap because the data is stored in an arbitrary order.

### Adding options

To add an option, a few things have to be done:

1. Add the widget of choice to the _optionsDialog.ui_
	file in designer mode. You do not need to _go to slot…_

2. Choose an appropriate, obvious and short name for
	the option/setting and rename the widget something similar. 
    [For example: the option for the size of the bins is called _binsize_
	and the corresponding widget (a LineEdit) is called  _binSizeLineEdit_.
	It should not be necessary to switch to the ui to look up the name.]

3. In _optionsDialog.cpp_,
	add the appropriate expression for saving the option to _setCurrentSettings()_.
    [For example: the binsize is set like this:
    > _currentSettings["binsize"]=ui->binSizeLineEdit->text().toInt();]

4. Also add an expression for resetting the option in _resetSettings()._

5. In _QCstmDQE.cpp,_ add an expression for setting the corresponding private variable in
	_on_apply_options()_.
	Or start some procedure.
    [For example: QCstmDQE has a variable called _windowW_
	that is used to locally fit the data (smoothing). This is set by the
	following expression:
    > _windowW=options.value("windowW");

    In the case of __binsize_, if it has changed, the binned data is immediately plotted again.]

6. Test.


## Works Cited

Dobbins. (1995). DQE(f) of four generations of
computed radiography acquisition devices. _Medical
Physics_.

Flynn, & Samei. (1999). Experimental
comparison of noise and resolution for 2k and 4k storage phosphor
radiography systems. _Medical Physics_.

Hoeschen, D. (2001). DQE of digital x-ray imaging
systems: a challenge. _SPIE_.


RELEASE_NOTES.md
## Release Notes


* 2.0.0
    * 2019/07/04 Freedom RC
    * ...
* 1.9.10
    * 2019/06/20
    * ...
* 1.9.9
    * 2019/04/26
    * ...
* 1.9.8
    * 2018/11/01
    * hotfix-dls-Jan2019-demo
        * ...
    * ...

...

* 1.6.1
    * 2017/08/14
    * Autosave bugfix and minor improvements
* 1.6.0
    * 2017/08/09
    * Henk’s Double Counter workaround in LEON software
* 1.5.0
    * 2017/08/01
    * New major chip feature. Test pulses works!
* 1.4.1
    * 2017/07/21
    * Minor bug fix - Threshold scan - number of frames = 1
* 1.4.0
    * 2017/07/21
    * New feature - Threshold scan
        * To be extended for spectroscopic units
* 1.3.1
    * 2017/07/18
    * Save both TIFFs at once
    * Threshold scan in development
    * 2017/06/13
    * Minor bugfix - save as text file with all thresholds
* 1.3.0
    * 2017/05/30
    * SpidrMpx3Lib big update
    * Speed tweaking
    * Added alpha feature - threshold scan
* 1.2.4
    * 2017/05/21
    * Minor bug fix - keyboard shortcut fix
* 1.2.3
    * 2017/05/17
    * Major bug fix - SpidrMpx3Lib fix (crash on connection)
* 1.2.2
    * 2017/05/12
    * Minor bug fix - keyboard shortcut
* 1.2.1
    * 2017/05/11
    * Various keyboard shortcuts for GUI automation added
* 1.2.0
    * 2017/04/24
    * Developer mode added
    * Autosave improved
    * Icon fix
    * Updated Author information
* 1.1.0
    * 2017/03/22
    * Equalisation bug fix
* 1.0.0
    * 2017/03/20
    * First release version


SpidrMpx3Eq/DACs.h
/**
 * John Idarraga <idarraga@cern.ch>
 * Nikhef, 2014.
 */

#ifndef DACS_H
#define DACS_H

#include "ui_mpx3gui.h"

#include "mpx3gui.h"
#include "mpx3defs.h"

#include <QDialog>
#include <QThread>
#include <QJsonObject>

#include <map>
#include <set>
#include <vector>
#include <string>
#include <iostream>
#include <fstream>
using namespace std;

#define 	__nDACs_MPX3RX		27
#define		__default_DACs_filename "mpx3_defaultDACs.json"
#define		__N_RETRY_ORIGINAL_SETTING	3

class SpidrController;
class SpidrDaq;
class QCustomPlot;
class QCPGraph;


// Taken from LEON Software.  Henk.
typedef struct dac_s
{
	int         code;
	const char *name;
	int         offset;
	int         size;
	int         dflt;
} dac_t;

// Tables with descriptions of DACs in the DACs register
// according to device type

// NB: with the 'offset' values given here (taken from the Medipix3.1 manual),
// a bit array of length 256 (32 bytes) should be used !
// (which is the correct length of the MPX3 DACs register;
//  existing code had 'Threshold[0]' offset at 0, etc, requiring an array
//  length value of 210, when filling it using 'bitarray' functions,
//  which is a bit strange...)

static const dac_t MPX3RX_DAC_TABLE[MPX3RX_DAC_COUNT] =
{
		{  1, "Threshold0",     30, 9, (1<<9)/2 },
		{  2, "Threshold1",     39, 9, (1<<9)/2 },
		{  3, "Threshold2",     48, 9, (1<<9)/2 },
		{  4, "Threshold3",     57, 9, (1<<9)/2 },
		{  5, "Threshold4",     66, 9, (1<<9)/2 },
		{  6, "Threshold5",     75, 9, (1<<9)/2 },
		{  7, "Threshold6",     84, 9, (1<<9)/2 },
		{  8, "Threshold7",     93, 9, (1<<9)/2 },
		{  9, "I_Preamp",          102, 8, (1<<8)/2 },
		{ 10, "I_Ikrum",           110, 8, (1<<8)/2 },
		{ 11, "I_Shaper",          118, 8, (1<<8)/2 },
		{ 12, "I_Disc",            126, 8, (1<<8)/2 },
		{ 13, "I_Disc_LS",         134, 8, (1<<8)/2 },
		{ 14, "I_Shaper_Test",     142, 8, (1<<8)/2 },
		{ 15, "I_DAC_DiscL",       150, 8, (1<<8)/2 },
		{ 30, "I_DAC_test",        158, 8, (1<<8)/2 },
		{ 31, "I_DAC_DiscH",       166, 8, (1<<8)/2 },
		{ 16, "I_Delay",           174, 8, (1<<8)/2 },
		{ 17, "I_TP_BufferIn",     182, 8, (1<<8)/2 },
		{ 18, "I_TP_BufferOut",    190, 8, (1<<8)/2 },
		{ 19, "V_Rpz",             198, 8, (1<<8)/2 },
		{ 20, "V_Gnd",             206, 8, (1<<8)/2 },
		{ 21, "V_Tp_ref",          214, 8, (1<<8)/2 },
		{ 22, "V_Fbk",             222, 8, (1<<8)/2 },
		{ 23, "V_Cas",             230, 8, (1<<8)/2 },
		{ 24, "V_Tp_refA",         238, 9, (1<<9)/2 },
		{ 25, "V_Tp_refB",         247, 9, (1<<9)/2 }
};

static const QColor COLOR_TABLE[] = {
		Qt::red, // 1
		Qt::black,
		Qt::darkRed,
		Qt::green,
		Qt::darkGreen,
		Qt::blue,
		Qt::darkBlue,
		Qt::cyan,
		Qt::darkCyan,
		Qt::magenta, // 10
		Qt::darkMagenta,
		Qt::yellow,
		Qt::darkYellow,
		QColor( "darkorange" ),
		QColor( "purple" ),
		QColor( "khaki" ),
		QColor( "gold" ),
		QColor( "dodgerblue" ), // 18
		QColor( "light gray" ),
		QColor( "medium gray" ),
		QColor( "red" ),
		QColor( "green" ),
		QColor( "blue" ),
		QColor( "cyan" ),
		QColor( "magenta" ),
		QColor( "yellow" ),
		QColor( "dark yellow" ) // 27
};


class QSpinBox;
class QSlider;
class QLabel;
class QCheckBox;
class SenseDACsThread;
class ScanDACsThread;
class QSignalMapper;
class ModuleConnection;

class SignalSlotMapping : public QObject {

	Q_OBJECT

public:
	explicit SignalSlotMapping(){};
	~SignalSlotMapping(){};

	int index;
	int value;

};

class DACs : public QObject {

	Q_OBJECT

public:

	explicit DACs();
	explicit DACs(QApplication * coreApp, Ui::Mpx3GUI * );
	~DACs();
	void PopulateDACValues();

	bool ReadDACsFile(string);
	bool WriteDACsFile(string);


public:

	Ui::Mpx3GUI * GetUI() { return _ui; };
	QSpinBox ** GetSpinBoxList() { return _dacSpinBoxes; };
	QSlider ** GetSliderList() { return _dacSliders; };
	QLabel ** GetLabelsList() { return _dacVLabels ; };
	QCheckBox ** GetCheckBoxList() { return _dacCheckBoxes; };
	int GetDeviceIndex() { return _deviceIndex; };
	int GetNSamples() { return _nSamples; };
	int GetScanStep() { return _scanStep; };
	QCPGraph * GetGraph(int idx);
	QCustomPlot * GetQCustomPlotPtr() { return _dacScanPlot; };
	void SetMpx3GUI(Mpx3GUI * p) { _mpx3gui = p; };
	void getConfig();
	void setConfig();

private:

	void FillWidgetVectors();
	void SetLimits();
	/*all the configuration bits*/
	QJsonObject configJson;

	// Connectivity between modules
	Mpx3GUI * _mpx3gui;

	QApplication * _coreApp;
	Ui::Mpx3GUI * _ui;

	QCustomPlot * _dacScanPlot;
	// Currently active graph
	QCPGraph *_graph;

	//
	SenseDACsThread * _senseThread;
	ScanDACsThread * _scanThread;

	// Vectors of Widgets
	QSpinBox  * _dacSpinBoxes[MPX3RX_DAC_COUNT];
	QSlider   * _dacSliders[MPX3RX_DAC_COUNT];
	QLabel    * _dacVLabels[MPX3RX_DAC_COUNT];
	QLabel    * _dacLabels[MPX3RX_DAC_COUNT];
	QCheckBox * _dacCheckBoxes[MPX3RX_DAC_COUNT];


	// Keep track of DAC values
	int _dacVals[MPX3RX_DAC_COUNT];

	// Scan
	int _scanStep;
	// Current device Id
	int _deviceIndex;
	// Samples
	int _nSamples;

	// In case only a subset of the DACs are selected
	//  to produce the scan, keep track of the id's
	map<int, int> _plotIdxMap;
	int _plotIdxCntr;

	QSignalMapper * _signalMapperSliderSpinBoxConn;
	QSignalMapper * _signalMapperSlider;
	QSignalMapper * _signalMapperSpinBox;


private slots:

void setTextWithIdx(QString,int);
void UncheckAllDACs();
void CheckAllDACs();
void setValueDAC(int);
void StartDACScan();
void SetupSignalsAndSlots();
void FromSpinBoxUpdateSlider(int);
void FromSliderUpdateSpinBox(int);
void SenseDACs();
void ChangeDeviceIndex(int);
void ChangeNSamples(int);
void ChangeScanStep(int);
void addData(int, int, double);
void addData(int);
void scanFinished();
void slideAndSpin(int, int);
void openWriteMenu();
void ConnectionStatusChanged();


};

class SenseDACsThread : public QThread {

	Q_OBJECT

public:

	explicit SenseDACsThread (int devIndx, DACs * dacs, SpidrController * sc);

private:

	SpidrController * _spidrcontrol;
	DACs * _dacs;
	int _deviceIndex;
	// IP source address (SPIDR network interface)
	int _srcAddr;

	void run();

signals:

	// These are used in the parent class as a signal to thread-safe feed
	//  widgets in the ui
	void progress(int);
	void fillText(QString);

};

class ScanDACsThread : public QThread {

	Q_OBJECT

public:

	explicit ScanDACsThread (int devIndx, DACs * dacs, SpidrController * sc);

private:

	SpidrController * _spidrcontrol;
	DACs * _dacs;
	int _deviceIndex;
	// IP source address (SPIDR network interface)
	int _srcAddr;

	void run();

signals:

	// These are used in the parent class as a signal to thread-safe feed
	//  widgets in the ui
	void progress(int);
	void fillText(QString);
	void fillTextWithIdx(QString, int);
	void addData(int, int, double);
	void addData(int);
	void scanFinished();
	void slideAndSpin(int, int);

};


#endif

